pkg/slayers/path/scion/info_hop_setter_lemmas.gobra

```diff
@@ -0,0 +1,519 @@
+}
+
+// CombineBytesFromInfoFields combines the at most three infofields of a packet into a single slice of bytes.
+ghost
+requires 0 < p
+requires 0 < numInf && numInf <= 3
+requires 0 <= offset
+requires offset + numInf*path.InfoLen <= len(raw)
+requires acc(sl.AbsSlice_Bytes(raw[:offset], 0, offset), p)
+requires acc(sl.AbsSlice_Bytes(raw[offset:offset+path.InfoLen], 0, path.InfoLen), p)
+requires 1 < numInf ==> acc(sl.AbsSlice_Bytes(raw[offset+path.InfoLen:offset+2*path.InfoLen], 0, path.InfoLen), p)
+requires 2 < numInf ==> acc(sl.AbsSlice_Bytes(raw[offset+2*path.InfoLen:offset+3*path.InfoLen], 0, path.InfoLen), p)
+requires acc(sl.AbsSlice_Bytes(raw[offset+numInf*path.InfoLen:], 0, len(raw[offset+numInf*path.InfoLen:])), p)
+ensures  acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), p)
+decreases
+func CombineBytesFromInfoFields(raw []byte, offset int, numInf int, p perm) {
+	sl.Unslice_Bytes(raw, offset+numInf*path.InfoLen, len(raw), p)
+	if(numInf > 2) {
+		sl.Unslice_Bytes(raw, offset+2*path.InfoLen, offset+3*path.InfoLen, p)
+		sl.CombineAtIndex_Bytes(raw, offset+2*path.InfoLen, len(raw), offset+3*path.InfoLen, p)
+	}
+	if(numInf > 1) {
+		sl.Unslice_Bytes(raw, offset+path.InfoLen, offset+2*path.InfoLen, p)
+		sl.CombineAtIndex_Bytes(raw, offset+path.InfoLen, len(raw), offset+2*path.InfoLen, p)
+
+	}
+	sl.Unslice_Bytes(raw, offset, offset+path.InfoLen, p)
+	sl.Unslice_Bytes(raw, 0, offset, p)
+	sl.CombineAtIndex_Bytes(raw, offset, len(raw), offset+path.InfoLen, p)
+	sl.CombineAtIndex_Bytes(raw, 0, len(raw), offset, p)
+}
+
+// CurrSegWithInfo depends only on the hopfield segment rather than the entire raw bytes of the packet.
+// This simplifies proving changes within a segment after updates to the packet's raw bytes.
+ghost
+opaque
+requires 0 < segLen
+requires 0 <= currHfIdx && currHfIdx <= segLen
+requires segLen*path.HopLen == len(hopfields)
+requires acc(sl.AbsSlice_Bytes(hopfields, 0, len(hopfields)), R56)
+decreases
+pure func CurrSegWithInfo(hopfields []byte, currHfIdx int, segLen int, inf path.IntermediateAbsInfoField) io.IO_seg3 {
+	return segment(hopfields, 0, currHfIdx, inf.AInfo, inf.UInfo, inf.ConsDir, inf.Peer, segLen)
+}
+
+// LeftSegWithInfo depends only on the hopfield segment rather than the entire raw bytes of the packet.
+// Depending on which segment of the packet is represented by the left segment (currInfIdx), it requires
+// permissions to a different slice of the packet's bytes.
+ghost
+opaque
+requires 0 < seg1Len
+requires 0 <= seg2Len
+requires 0 <= seg3Len
+requires currInfIdx == 1 && seg2Len > 0 ==>
+	inf != none[path.IntermediateAbsInfoField] &&
+	acc(sl.AbsSlice_Bytes(hopfields, 0, len(hopfields)), R56) &&
+	seg2Len*path.HopLen == len(hopfields)
+requires currInfIdx == 2 && seg2Len > 0 && seg3Len > 0 ==>
+	inf != none[path.IntermediateAbsInfoField] &&
+	acc(sl.AbsSlice_Bytes(hopfields, 0, len(hopfields)), R56) &&
+	seg3Len*path.HopLen == len(hopfields)
```

Comment:
if we introduce the structure with the segLens, we could introduce a function `ExpectedHopFieldsLen` that performs these computations, instead of putting these large expressions in the preconditions. This would simplify some of these lemmas